be/src/exprs/function_call_expr.cpp

```diff
@@ -177,52 +177,53 @@ StatusOr<ColumnPtr> VectorizedFunctionCallExpr::evaluate_checked(starrocks::Expr
 }
 
 bool VectorizedFunctionCallExpr::ngram_bloom_filter(ExprContext* context, const BloomFilter* bf,
-                                                    size_t index_gram_num) const {
+                                                    const NgramBloomFilterReaderOptions& reader_options) const {
     FunctionContext* fn_ctx = context->fn_context(_fn_context_index);
-    std::vector<Slice>& ngram_set = fn_ctx->get_ngram_set();
-
-    const auto& gram_num_column = fn_ctx->get_constant_column(2);
-    // case like ngram_search(col,"needle", 5) when col has a 4gram bloom filter, don't use this index
-    if (gram_num_column != nullptr) {
-        size_t predicate_gram_num = ColumnHelper::get_const_value<TYPE_INT>(gram_num_column);
-        if (index_gram_num != predicate_gram_num) {
-            return true;
-        }
-    }
-
-    if (UNLIKELY(ngram_set.size() == 0)) {
-        Slice needle;
-        if (_fn_desc->name == "like" || _fn_desc->name == "regex") {
-            // not implemented right now
-            return true;
+    std::unique_ptr<NgramBloomFilterState>& ngram_state = fn_ctx->get_ngram_state();
+
+    // initialize ngram_state: determine whether this index useful or not, split needle into ngram_set if useful
+    if (ngram_state == nullptr) {
+        ngram_state = std::make_unique<NgramBloomFilterState>();
+        std::vector<Slice>& ngram_set = ngram_state->ngram_set;
+        bool index_useful;
+        if (_fn_desc->name == "LIKE") {
```

Comment:
Why change from like to LIKE?

Reply:
[60010, 'LIKE', 'BOOLEAN', ['VARCHAR', 'VARCHAR'], 'LikePredicate::like', 'LikePredicate::like_prepare',  which cause function name is "LIKE"